Meta: Add GDB pretty printers

gunnarbeutner opened the pull request:
This adds support for GDB pretty printers and uses them by default for `Meta/serenity.sh gdb` and `Meta/debug-kernel.sh`.

TODO:

- [x] Remove `-ggdb3` or make it configurable.

![Screenshot from 2021-04-21 15-10-47](https://user-images.githubusercontent.com/388571/115559237-cf8e7e80-a2b3-11eb-986f-2c9c17a888fa.png)

linusg commented on Meta/serenity_gdb.py:
```diff
@@ -0,0 +1,263 @@
+import re
+
+
+class AKAtomic(object):
```
new style classes have been the default in Python for a long time, none of these need to explicitly inherit from object.

linusg commented on Meta/serenity_gdb.py:
```diff
@@ -0,0 +1,263 @@
+
+    def to_string(self):
+        if int(self.val["m_impl"]["m_bits"]["m_value"]) == 0:
+            return "\"\""
```
```suggestion
            return '""'
```

Same below.

linusg commented on Meta/serenity_gdb.py:
```diff
@@ -0,0 +1,263 @@
+    # Trying to access a variable-length field seems to fail with
+    # Python Exception <class 'gdb.error'> value requires 4294967296 bytes, which is more than max-value-size
+    # This works around that issue.
+    return gdb.parse_and_eval("*(%s*)((%s*)%s)->%s" % (type, val.type.name, int(val.address), member))
```
We can assume modern Python >= 3.6 (3.5 and below had EOL already), please use f-strings. Same below.

bgianfo commented on Meta/serenity_gdb.py:
```diff
@@ -0,0 +1,263 @@
+gdb.printing.register_pretty_printer(None, SerenityPrettyPrinterLocator(), replace=True)
+
+
+class FindThreadCmd(gdb.Command):
```
Can you write python help text, so it will show up when someone asks for help of this command in gdb? 

Reference: https://sourceware.org/gdb/onlinedocs/gdb/Commands-In-Python.html

> The help text for the new command is taken from the Python documentation string for the command’s class, if there is one. If no documentation string is provided, the default value “This command is not documented.” is used.

bgianfo reviewed (changes requested):
Looks good in general, just a few small nits from me.

gunnarbeutner commented:
Docs are still missing. Will add those in a bit. Thanks for the reviews!

bgianfo commented on Kernel/CMakeLists.txt:
```diff
@@ -1,4 +1,4 @@
-add_compile_options(-Os)
+add_compile_options(-Os -ggdb3)
```
Just a reminder, still  probably need to make this conditional based on; 

```
commit a731ccd4a0bd55dd1a7a320c4c6e9bc383cb2618
Author: Andreas Kling <kling@serenityos.org>
Date:   Sat Feb 22 18:39:22 2020 +0100

    Kernel: Build without debugging symbols by default

    Compiling with -g adds roughly 30% to kernel build times. Anyone who
    wants this can turn it on locally instead.

commit 46c60fd451e478ee0bc834de7ff129d85d437bb3
Author: Brian Gianforcaro <b.gianfo@gmail.com>
Date:   Sun Jan 12 22:50:47 2020 -0800

    Debugging: Add kernel debugging support

    Introduce the 'debug-kernel' script to allow developers to
    quickly attach a debugger to the QEMU debug remote. The
    setting (-s) is already enabled by ./run today when using
    QEMU for virtualisation.

    If the system is running under QEMU, the debugger
    will break in when the script is run. If you add
    the -S option to QEMU it will wait for the debugger
    to connect before booting the kernel. This allows
    you to debug the init/boot process.

    Personally I use cgdb instead of gdb, so I opted
    to make the debugger used by the script customizable
    via an environment variable.

    This change also adds -g3 to the kernel build so that
    rich debug symbols are available in the kernel binary.
```

gunnarbeutner replied:
Ah yes, I thought I had removed that for now. Turns out I re-added it to test my other changes. :+1:

bgianfo replied:
It would be interesting to know how much of a regression this really is these days, considering when @awesomekling  reverted the change we weren't even compiling with optimizations enabled and were probably using a 9 series gcc.

bgianfo replied:
Here's the current numbers from my machine.

Without;  
```
Without: 
$ time ninja Kernel
[0/2] Re-checking globbed directories...
[259/259] Linking CXX executable Kernel/Kernel

real	0m47.773s
user	5m57.104s
sys	0m11.670s
```
With `-gddb3`:
```
$ time ninja Kernel
[0/2] Re-checking globbed directories...
[259/259] Linking CXX executable Kernel/Kernel

real	1m2.625s
user	7m44.211s
sys	0m16.173s
```

gunnarbeutner replied:
Sounds like we'd probably want a cmake option for it. :)

gunnarbeutner commented:
* Removed `-ggdb3`.
* Added docs:

```
(gdb) help find_thread

    Find SerenityOS thread for the specified TID.
    find_thread TID
    
(gdb) 
```

linusg commented on Meta/serenity_gdb.py:
```diff
@@ -0,0 +1,274 @@
+        else:
+            elements = get_field_unalloced(self.val, "m_inline_buffer_storage", inner_type_ptr)
+
+        return [("[%d]" % (i,), elements[i]) for i in range(vec_len)]
```
```suggestion
        return [(f"[{i}]", elements[i]) for i in range(vec_len)]
```

gunnarbeutner replied:
Oops, fixed.

linusg reviewed (commented):
Very nice how gdb integrates with Python for pretty-printing - didn't know this was possible :)
